# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1379957462 14400
#      Mon Sep 23 13:31:02 2013 -0400
# Node ID cc13924a4266fb0359f59fabdce11071e6051d48
# Parent  177147bf7b55e53c19e2c4340c277f7b9a9162af
snapshot 3.7.7

* configure.ac (OCTAVE_VERSION): Bump to 3.7.7.

diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -14,23 +14,23 @@ dnl Copyright (C) 1993-2012 John W. Eato
 ### FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 ### for more details.
 ### 
 ### You should have received a copy of the GNU General Public License
 ### along with Octave; see the file COPYING.  If not, see
 ### <http://www.gnu.org/licenses/>.
 
 AC_PREREQ([2.62])
-AC_INIT([GNU Octave], [3.7.6+], [http://octave.org/bugs.html], [octave])
+AC_INIT([GNU Octave], [3.7.7], [http://octave.org/bugs.html], [octave])
 
 dnl PACKAGE_VERSION is set by the AC_INIT VERSION arg
 OCTAVE_VERSION="$PACKAGE_VERSION"
 OCTAVE_API_VERSION_NUMBER="48"
 OCTAVE_API_VERSION="api-v$OCTAVE_API_VERSION_NUMBER+"
-OCTAVE_RELEASE_DATE="2013-08-20"
+OCTAVE_RELEASE_DATE="2013-09-23"
 OCTAVE_COPYRIGHT="Copyright (C) 2013 John W. Eaton and others."
 AC_SUBST(OCTAVE_VERSION)
 AC_SUBST(OCTAVE_API_VERSION_NUMBER)
 AC_SUBST(OCTAVE_API_VERSION)
 AC_SUBST(OCTAVE_RELEASE_DATE)
 AC_SUBST(OCTAVE_COPYRIGHT)
 
 dnl FIXME: We should auto-insert the Mercurial changeset ID into the
diff --git a/libinterp/corefcn/oct-stream.cc b/libinterp/corefcn/oct-stream.cc
--- a/libinterp/corefcn/oct-stream.cc
+++ b/libinterp/corefcn/oct-stream.cc
@@ -2975,31 +2975,57 @@ convert_and_copy (std::list<void *>& inp
 
   octave_idx_type j = 0;
 
   for (std::list<void *>::const_iterator it = input_buf_list.begin ();
        it != input_buf_list.end (); it++)
     {
       SRC_T *data = static_cast<SRC_T *> (*it);
 
-      for (octave_idx_type i = 0; i < input_buf_elts && j < elts_read; i++, j++)
+      if (swap || do_float_fmt_conv)
         {
-          if (swap)
-            swap_bytes<sizeof (SRC_T)> (&data[i]);
-          else if (do_float_fmt_conv)
-            do_float_format_conversion (&data[i], sizeof (SRC_T),
-                                        1, from_flt_fmt,
-                                        oct_mach_info::float_format ());
-
-          dst_elt_type tmp (data[i]);
-
-          if (do_NA_conv && __lo_ieee_is_old_NA (tmp))
-            tmp = __lo_ieee_replace_old_NA (tmp);
-
-          conv_data[j] = tmp;
+          for (octave_idx_type i = 0; i < input_buf_elts && j < elts_read;
+               i++, j++)
+            {
+              if (swap)
+                swap_bytes<sizeof (SRC_T)> (&data[i]);
+              else if (do_float_fmt_conv)
+                do_float_format_conversion (&data[i], sizeof (SRC_T),
+                                            1, from_flt_fmt,
+                                            oct_mach_info::float_format ());
+
+              dst_elt_type tmp (data[i]);
+
+              if (do_NA_conv && __lo_ieee_is_old_NA (tmp))
+                tmp = __lo_ieee_replace_old_NA (tmp);
+
+              conv_data[j] = tmp;
+            }
+        }
+      else
+        {
+          if (do_NA_conv)
+            {
+              for (octave_idx_type i = 0; i < input_buf_elts && j < elts_read;
+                   i++, j++)
+                {
+                  dst_elt_type tmp (data[i]);
+
+                  if (__lo_ieee_is_old_NA (tmp))
+                    tmp = __lo_ieee_replace_old_NA (tmp);
+
+                  conv_data[j] = tmp;
+                }
+            }
+          else
+            {
+              for (octave_idx_type i = 0; i < input_buf_elts && j < elts_read;
+                   i++, j++)
+                conv_data[j] = data[i];
+            }
         }
 
       delete [] data;
     }
 
   input_buf_list.clear ();
 
   for (octave_idx_type i = elts_read; i < nr * nc; i++)
@@ -3012,17 +3038,16 @@ typedef octave_value (*conv_fptr)
   (std::list<void *>& input_buf_list, octave_idx_type input_buf_elts,
    octave_idx_type elts_read, octave_idx_type nr, octave_idx_type nc,
    bool swap, bool do_float_fmt_conv, bool do_NA_conv,
    oct_mach_info::float_format from_flt_fmt);
 
 #define TABLE_ELT(T, U, V, W) \
   conv_fptr_table[oct_data_conv::T][oct_data_conv::U] = convert_and_copy<V, W>
 
-#undef FILL_TABLE_ROW
 #define FILL_TABLE_ROW(T, V) \
   TABLE_ELT (T, dt_int8, V, int8NDArray); \
   TABLE_ELT (T, dt_uint8, V, uint8NDArray); \
   TABLE_ELT (T, dt_int16, V, int16NDArray); \
   TABLE_ELT (T, dt_uint16, V, uint16NDArray); \
   TABLE_ELT (T, dt_int32, V, int32NDArray); \
   TABLE_ELT (T, dt_uint32, V, uint32NDArray); \
   TABLE_ELT (T, dt_int64, V, int64NDArray); \
